### PR TITLE
Projector.service user slide

### PR DESCRIPTION
**Pull request by r-peschke**
This PR is based on master

### CODE REVIEW

**Comment by ostcar** [comment]
Please have a look to my comments in #206 and make a first review by yourself. Please request my review afterwards

**Review by ostcar — CHANGES_REQUESTED** [review]
Two open questions:
* Errors for the client
* title information

File: internal/projector/slide/user.go [hunk 1]
```diff
@@ -72,40 +95,28 @@ func (u *dbUser) StringMeetingDependent(meetingID int) string {
...
+		if err != nil {
+			return nil, nil, fmt.Errorf("getting user id: %w", err)
+		}
+
+		user, keys, err := newUser(ctx, ds, id, p7on.MeetingID)
+		if err != nil {
+			return nil, nil, fmt.Errorf("loading user: %w", err)
+		}
+
+		responseValue, err := json.Marshal(map[string]interface{}{"user": user.GetUserRepresentation(p7on.MeetingID)})
```
Comment (ostcar) [inline internal/projector/slide/user.go:114]: I don't like this. I would propose that you define a struct:
```go
out := struct{
  User string `json:"user"`
}{
  user.GetUserRepresentation(p7on.MeetingID),
}
responseValue, err := json.Marshal(out)
```

I know this is more to write, but easier to read.

But if makes you happy, you could tolerate your version at this place.

**Review by ostcar — CHANGES_REQUESTED** [review]

File: internal/projector/projector.go [hunk 2]
```diff
@@ -54,7 +54,8 @@ func Register(ds Datastore, slides *SlideStore) {
 
 		parts := strings.SplitN(fqfield, "/", 3)
 		if len(parts) != 3 {
-			return nil, fmt.Errorf("invalid key %s, expected two '/'", fqfield)
+			err = fmt.Errorf("invalid key %s, expected two '/'", fqfield)
+			return bs, err
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply internal/projector/projector.go:58]: Done, missed it in rebasing

File: internal/projector/slide/list_of_speakers.go [hunk 3]
```diff
@@ -107,7 +107,7 @@ func renderListOfSpeakers(ctx context.Context, ds projector.Datastore, losFQID s
 		addKeys = append(addKeys, keys...)
 
 		s := outputSpeaker{
-			User:         user.StringMeetingDependent(meetingID),
+			User:         user.GetUserRepresentation(meetingID),
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply internal/projector/slide/list_of_speakers.go:110]: Done

File: internal/projector/slide/user.go [hunk 4]
```diff
@@ -20,6 +20,7 @@ type dbUser struct {
 	DefaultLevel string `json:"default_structure_level"`
 }
 
+// newUser gets the user from datastore and return the user as dbUser struct together with keys and error
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply internal/projector/slide/user.go:23]: done

File: internal/projector/slide/user.go [hunk 5]
```diff
@@ -43,19 +44,41 @@ func newUser(ctx context.Context, ds datastore.Getter, id, meetingID int) (*dbUs
...
 
+	if u.FirstName == "" && u.LastName == "" && u.Username == "" {
+		return nil, nil, fmt.Errorf("neither firstName, lastName nor username found")
+	}
 	return &u, keys, nil
 }
 
-// StringMeetingDependent gets the instance representation of the user, which is meeting dependent with the structur_level
-func (u *dbUser) StringMeetingDependent(meetingID int) string {
+// getUserRepresentation returns the meeting-dependent string for the given user
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply internal/projector/slide/user.go:61]: done

File: internal/projector/slide/user.go [hunk 6]
```diff
@@ -43,19 +44,41 @@ func newUser(ctx context.Context, ds datastore.Getter, id, meetingID int) (*dbUs
...
+func (u *dbUser) GetUserRepresentation(meetingID int) string {
+	parts := []string{u.getUserShortName()}
+	level := u.GetUserStructureLevel(meetingID)
+	if level != "" {
+		parts = append(parts, fmt.Sprintf("(%s)", level))
+	}
+	return strings.Join(parts, " ")
+}
+
+func (u *dbUser) GetUserStructureLevel(meetingID int) string {
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply internal/projector/slide/user.go:71]: done

File: internal/projector/slide/user.go [hunk 7]
```diff
@@ -43,19 +44,41 @@ func newUser(ctx context.Context, ds datastore.Getter, id, meetingID int) (*dbUs
...
+	level := u.Level
+	if level == "" {
+		level = u.DefaultLevel
+	}
+	return level
+}
+
+// getUserShortName returns the short name as "title first_name last_name".
+// Without first_name and last_name, uses username instead
+func (u *dbUser) getUserShortName() string {
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply internal/projector/slide/user.go:81]: Changed to capital letter for export

File: internal/projector/slide/user.go (hunk 1 above)
Comment (r-peschke) [reply internal/projector/slide/user.go:114]: changed

File: internal/projector/slide/user.go [hunk 8]
```diff
@@ -43,19 +45,43 @@ func newUser(ctx context.Context, ds datastore.Getter, id, meetingID int) (*dbUs
...
+	if u.FirstName == "" && u.LastName == "" && u.Username == "" {
+		return nil, nil, fmt.Errorf("neither firstName, lastName nor username found")
+	}
 	return &u, keys, nil
 }
 
-// StringMeetingDependent gets the instance representation of the user, which is meeting dependent with the structur_level
-func (u *dbUser) StringMeetingDependent(meetingID int) string {
+// UserRepresentation returns the meeting-dependent string for the given user.
+func (u *dbUser) UserRepresentation(meetingID int) string {
```
Comment (ostcar) [inline internal/projector/slide/user.go:63]: Only a proposal:

I would have written the function in this way:

```go
name := u.UserShortName()
level := u.UserStructureLevel(meetingID)
if level == "" {
  return name
}
return fmt.Sprintf("%s (%s)", name, level)
```

It does not seem to be necessary to create a slice and join it at the end.

File: internal/projector/slide/user.go [hunk 9]
```diff
@@ -43,19 +45,43 @@ func newUser(ctx context.Context, ds datastore.Getter, id, meetingID int) (*dbUs
...
+	}
+	return strings.Join(parts, " ")
+}
+
+// UserStructureLevel returns in first place the meeting specific level,
+// otherwise the default level.
+func (u *dbUser) UserStructureLevel(meetingID int) string {
+	level := u.Level
+	if level == "" {
+		level = u.DefaultLevel
```
Comment (ostcar) [inline internal/projector/slide/user.go:77]: Only a proposal:

It is not unusual in go to have more then one return statement in a function. The rule from other languages, that there should only be one return statement at the end of a function does not apply to go. It solved the problem with the `defer` statement.

With this in mind, I would have wrote this function like this:

```go
if u.Level == "" {
  return u.DefaultLevel
}
return u.Level
```

Also: The attribute `meetingID` is not used in this function. Can it be removed? If not, maybe add a comment why it is necessary (for example to satisfy an interface?

**Review by ostcar — APPROVED** [review]
Only two proposals. If you like, you can merge this

File: internal/projector/slide/user.go (hunk 8 above)
Comment (r-peschke) [reply internal/projector/slide/user.go:63]: Done, this is really better

**Pull request merged**